stablehlo/dialect/Serialization.cpp

```diff
@@ -0,0 +1,89 @@
+/* Copyright 2023 The StableHLO Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "stablehlo/dialect/Serialization.h"
+
+#include <sstream>
+
+#include "llvm/Support/raw_ostream.h"
+#include "mlir/Bytecode/BytecodeWriter.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Support/LogicalResult.h"
+#include "stablehlo/dialect/Version.h"
+#include "stablehlo/dialect/VhloOps.h"
+#include "stablehlo/transforms/Passes.h"
+
+namespace mlir {
+namespace stablehlo {
+
+LogicalResult serializePortableArtifact(ModuleOp module,
+                                        std::string const& targetVersion,
+                                        raw_ostream& os) {
+  MLIRContext* context = module.getContext();
+
+  // Convert StableHLO --> VHLO. Will fail if entire program is not StableHLO.
+  {
+    mlir::PassManager pm(context);
+    pm.addPass(mlir::stablehlo::createStablehloLegalizeToVhloPass());
+    if (!mlir::succeeded(pm.run(module))) {
+      return failure();
+    }
+  }
+
+  // Convert VHLO --> VHLO(version x.y.z).
+  // Doing separately for now since we need to improve error messaging around
+  // target version failures.
+  {
+    mlir::PassManager pm(context);
+    pm.addPass(mlir::stablehlo::createVhloToVersionPass({targetVersion}));
+    if (!mlir::succeeded(pm.run(module))) {
+      return failure();
+    }
+  }
+
+  // Writes the header VHLO_<serializerVersion>_<targetVersion>
+  std::string artifactHeader;
+  llvm::raw_string_ostream headerOs(artifactHeader);
+  headerOs << "VHLO_" << vhlo::Version::getCurrentVersion() << "_"
+           << vhlo::Version::fromString(targetVersion);
+  BytecodeWriterConfig config(artifactHeader);
+  writeBytecodeToFile(module, os, config);
+  return success();
+}
+
+mlir::OwningOpRef<mlir::ModuleOp> deserializePortableArtifact(
+    llvm::StringRef sourceStr, MLIRContext* context) {
+  // FIXME: Not sure if the lifetime is correct here.
```

Comment:
What are you not sure about? Let's discuss.

Reply:
Figured this out, forgot to delete the fixme. Wasn't sure on lifetime of an owning ref back when this API didn't return the owning ref. Now this is in-line with some upstream APIs.